src/module_3/module3_statistical_learning.md

```diff
@@ -0,0 +1,1368 @@
+
+    # Generate a mask for the upper triangle
+    mask = np.triu(np.ones_like(corr, dtype=bool))
+
+    # Set up the matplotlib figure
+    fig, ax = plt.subplots (figsize=(11, 9))
+
+    # Generate a custom diverging colormap
+    cmap = sns.diverging_palette (230, 20, as_cmap=True)
+
+    # Draw the heatmap with the mask and correct aspect ratio
+    sns.heatmap(corr, mask=mask, cmap=cmap, vmax=.3, center=0,# annot=True,
+            square=True, linewidths=.5, cbar_kws={"shrink": .5})
+    
+plot_correlation_matrix(df_orders, columns)
+```
+
+
+    
+![png](module3_statistical_learning_files/module3_statistical_learning_9_0.png)
+    
+
+
+## Create train and test sets
+
+Puntos a tener en cuenta
+- Hacer el split por order_id - No puede haber un order_id en train y test
+- Debemos tener en cuenta las fechas -  Ventas futuras de un mismo usuario se verán influenciadas por ventas pasadas (producto previamente en abandoned_cart o si un producto se ha comprado recientemente)
+
+
+```python
+# Before splitting the data we'll handle datetimes
+
+### Extract components of datetimes
+df_orders = df_orders.drop('created_at', axis=1) # I'll only keep order_date as they display almost the same info
+
+df_orders['order_year'] = df_orders['order_date'].dt.year
+df_orders['order_month'] = df_orders['order_date'].dt.month
+df_orders['order_day_of_week'] = df_orders['order_date'].dt.dayofweek 
+
+df_orders = df_orders.drop('order_date', axis=1)
+```
+
+
+```python
+def split_by_order_id(df, test_size=0.1, validation_size=0.1, random_state=42):
+    """
+    Splits a DataFrame into train, validation, and test sets based on unique order IDs.
+
+    Args:
+        df (pd.DataFrame): The input DataFrame.
+        test_size (float, optional): Proportion of the dataset to include in the test set.
+                                     Defaults to 0.2.
+        validation_size (float, optional): Proportion of the dataset to include in the validation set.
+                                           Defaults to 0.2.
+        random_state (int, optional): Controls the randomness of the split. 
+                                      Set for reproducible results. Defaults to None.
+
+    Returns:
+        tuple: A tuple containing three DataFrames: (train_df, validation_df, test_df)
+    """
+
+    # Get unique order IDs
+    unique_order_ids = df['order_id'].unique()
+
+    # Split order IDs into train and test groups 
+    train_order_ids, test_order_ids = train_test_split(unique_order_ids, 
+                                                       test_size=test_size, 
+                                                       random_state=random_state)
+
+    # Optionally, split train order IDs further into train and validation groups
+    if validation_size > 0:
+        train_order_ids, validation_order_ids = train_test_split(train_order_ids,
+                                                                 test_size=validation_size / (1 - test_size),
```

Comment:
this doesn't work, if `validation_size = 0.1` and `test_size = 0.1`, `validation_size / (1 - test_size) = 0.1/0.9 = 0.1111` so you are having a different size than 0.1 which is the one you passed in the parameter for the validation